File: lan-messenger-native/macos/LanMessenger/Core/Crypto/HistoryCrypto.swift
import Foundation
import CryptoKit

// Handles encryption/decryption of the local history file.
//
// Key derivation (does NOT use peer exchange):
//   history_key = HKDF-SHA256(
//       ikm  = raw_private_key_bytes (32 bytes),
//       salt = [] (empty),
//       info = "lan-messenger-history",
//       len  = 32
//   )
//
// AAD for all history operations: b"history-v1" (the literal 10 UTF-8 bytes)
//
// File format (JSON):
//   { "nonce": "<base64 12-byte nonce>", "ciphertext": "<base64 ct+tag>" }
//
// Inner plaintext is compact JSON (no extra whitespace) of the history dict.
enum HistoryCrypto {

    static let aad = Data("history-v1".utf8)

    static func historyKey(privateKey: Curve25519.KeyAgreement.PrivateKey) -> SymmetricKey {
        let rawPrivate = privateKey.rawRepresentation
        // HKDF with empty salt and info = "lan-messenger-history"
        return HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: rawPrivate),
            salt: Data(),
            info: Data("lan-messenger-history".utf8),
            outputByteCount: 32
        )
    }

    // Encrypt plaintext JSON bytes. Returns the outer file JSON string.
    static func encryptHistory(
        plaintext: Data,
        privateKey: Curve25519.KeyAgreement.PrivateKey
    ) throws -> String {
        let key = historyKey(privateKey: privateKey)
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(plaintext, using: key, nonce: nonce, authenticating: aad)
        let combined = sealed.ciphertext + sealed.tag
        let outer: [String: String] = [
            "nonce":      Data(nonce).base64EncodedString(),
            "ciphertext": combined.base64EncodedString(),
        ]
        let outData = try JSONSerialization.data(withJSONObject: outer)
        return String(data: outData, encoding: .utf8) ?? ""
    }

    // Decrypt from the outer file JSON string. Returns inner plaintext bytes.
    static func decryptHistory(
        fileJSON: String,
        privateKey: Curve25519.KeyAgreement.PrivateKey
    ) throws -> Data {
        guard let fileData = fileJSON.data(using: .utf8),
              let outer = try? JSONSerialization.jsonObject(with: fileData) as? [String: String],
              let nonceB64 = outer["nonce"],
              let ciphertextB64 = outer["ciphertext"],
              let nonceData = Data(base64Encoded: nonceB64), nonceData.count == 12,
              let combined = Data(base64Encoded: ciphertextB64), combined.count >= 16 else {
            throw SessionCryptoError.decryptionFailed
        }
        let key = historyKey(privateKey: privateKey)
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let ciphertext = combined.dropLast(16)
        let tag = combined.suffix(16)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        do {
            return try AES.GCM.open(box, using: key, authenticating: aad)
        } catch {
            throw SessionCryptoError.decryptionFailed
        }
    }
}

File: src/windows-native/LanMessenger/Core/Crypto/HistoryCrypto.cs
using NSec.Cryptography;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace LanMessenger.Core.Crypto;

// Handles encryption/decryption of the local history file.
//
// Key derivation (does NOT use peer exchange):
//   history_key = HKDF-SHA256(
//       ikm  = raw_private_key_bytes (32 bytes),
//       salt = [] (empty),
//       info = "lan-messenger-history" (UTF-8),
//       len  = 32
//   )
//
// AAD: b"history-v1" (the literal 10 UTF-8 bytes)
//
// File format (JSON): { "nonce": "<base64 12-byte nonce>", "ciphertext": "<base64 ct+tag>" }
//
// AES-GCM uses System.Security.Cryptography.AesGcm (BCL) rather than NSec's
// Aes256Gcm so it runs on CPUs without AES-NI (e.g. older Windows 10 machines).
public static class HistoryCrypto
{
    private static readonly HkdfSha256 _hkdf  = KeyDerivationAlgorithm.HkdfSha256;

    private static readonly byte[] _info  = Encoding.UTF8.GetBytes("lan-messenger-history");
    public  static readonly byte[] Aad    = Encoding.UTF8.GetBytes("history-v1");
    private static readonly byte[] _empty = [];

    private const int KeySize   = 32;
    private const int NonceSize = 12;
    private const int TagSize   = 16;

    // Derive the 32-byte history symmetric key from the private key bytes.
    public static byte[] HistoryKey(Key privateKey)
    {
        byte[] raw = privateKey.Export(KeyBlobFormat.RawPrivateKey);
        // Use BCL HKDF directly since we have raw IKM bytes here.
        return HKDF.DeriveKey(HashAlgorithmName.SHA256, raw, KeySize, salt: _empty, info: _info);
    }

    // Encrypt plaintext bytes. Returns the outer file JSON string.
    public static string EncryptHistory(byte[] plaintext, Key privateKey)
    {
        var keyBytes  = HistoryKey(privateKey);
        byte[] nonce      = new byte[NonceSize];
        byte[] ciphertext = new byte[plaintext.Length];
        byte[] tag        = new byte[TagSize];
        RandomNumberGenerator.Fill(nonce);

        using var aesGcm = new AesGcm(keyBytes, TagSize);
        aesGcm.Encrypt(nonce, plaintext, ciphertext, tag, Aad);

        byte[] ciphertextWithTag = new byte[ciphertext.Length + TagSize];
        ciphertext.CopyTo(ciphertextWithTag, 0);
        tag.CopyTo(ciphertextWithTag, ciphertext.Length);

        var outer = new Dictionary<string, string>
        {
            ["nonce"]      = Convert.ToBase64String(nonce),
            ["ciphertext"] = Convert.ToBase64String(ciphertextWithTag),
        };
        return JsonSerializer.Serialize(outer);
    }

    // Decrypt from the outer file JSON string. Returns inner plaintext bytes.
    public static byte[] DecryptHistory(string fileJson, Key privateKey)
    {
        using var doc = JsonDocument.Parse(fileJson);
        var root = doc.RootElement;

        byte[] nonce            = Convert.FromBase64String(root.GetProperty("nonce").GetString()!);
        byte[] ciphertextWithTag = Convert.FromBase64String(root.GetProperty("ciphertext").GetString()!);

        if (nonce.Length != NonceSize)          throw new CryptographicException("Invalid nonce length");
        if (ciphertextWithTag.Length < TagSize) throw new CryptographicException("Ciphertext too short");

        int    ciphertextLen = ciphertextWithTag.Length - TagSize;
        byte[] ciphertext    = ciphertextWithTag[..ciphertextLen];
        byte[] tag           = ciphertextWithTag[ciphertextLen..];
        byte[] plaintext     = new byte[ciphertextLen];

        var keyBytes = HistoryKey(privateKey);
        using var aesGcm = new AesGcm(keyBytes, TagSize);
        aesGcm.Decrypt(nonce, ciphertext, tag, plaintext, Aad);
        return plaintext;
    }
}
